### PR TITLE
Support versioning cluster and path level properties

### PR DESCRIPTION
**Pull request by cheng-chang**
This PR computes a hex encoded md5 hash of list of `key:value:source` strings as the version for cluster properties, and `path:key:value` for path properties. The version is lazily computed, cached, and updated.

### CODE REVIEW

**Comment by calvinjia** [comment]
@cheng-chang Based on the implementation, `version` seems a bit misleading because it implies there is an order. We are just computing a fingerprint for the properties.

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 1]
```diff
@@ -58,6 +59,11 @@
   /** Map of property sources. */
   private final ConcurrentHashMap<PropertyKey, Source> mSources = new ConcurrentHashMap<>();
 
+  private final Version mVersion = new Version(() -> keySet().stream()
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: What's the concurrency guarantee around this? Do all users of `AlluxioProperties` ensure thread safety?

**Comment by cheng-chang** [comment]
@calvinjia I think version is not a bad choice, it can indicate that different properties have different versions. Although I agree the current implementation doesn't support determining orders based on the versions.

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java (hunk 1 above)
Comment (cheng-chang) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: made `Version` thread safe.

**Comment by calvinjia** [comment]
@cheng-chang Instead of having a version field, could we just make it the equals implementation for the properties? I think we are only interested in whether or not the configs are equal, not if one is newer or older.

**Comment by cheng-chang** [comment]
@calvinjia these version values will be sent to and cached by clients periodically, clients rely on this to detect whether configurations have changed.

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (cheng-chang) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: removed final

File: core/common/src/main/java/alluxio/conf/AlluxioProperties.java [hunk 2]
```diff
@@ -146,6 +153,7 @@ public void merge(Map<?, ?> properties, Source source) {
       }
       put(propertyKey, value, source);
     }
+    mVersion.update();
```
[the remark this answers is not in the document]
Comment (cheng-chang) [reply core/common/src/main/java/alluxio/conf/AlluxioProperties.java]: renamed to `markOutdated`

File: core/common/src/main/java/alluxio/conf/Version.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  public Version(Supplier<Stream<byte[]>> properties) {
+    try {
+      mMD5 = MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+    mProperties = properties;
+    mShouldUpdate = new AtomicBoolean(true);
+    mVersion = "";
```
[the remark this answers is not in the document]
Comment (cheng-chang) [reply core/common/src/main/java/alluxio/conf/Version.java]: removed

**Comment by cheng-chang** [comment]
@aaudiber @calvinjia PTAL

**Comment by aaudiber** [comment]
LGTM modulo calling it a hash or fingerprint instead of a version

**Comment by cheng-chang** [comment]
@aaudiber I'll first merge this, then rename version to hash together with changes in https://github.com/Alluxio/alluxio/pull/8873.

**Comment by cheng-chang** [comment]
alluxio-bot, merge this please

**Pull request merged**